### PR TITLE
testing.parameterize can accept functions

### LINKED ISSUE #1337: testing.parameterize cannot take dictionary whose values are functions
testing.parameterize takes a list of dictionaries and add assign each element to attributes of the decorated class. But we cannot use functions as the value of the dictionaries as the functions will be registered as the methods of the decorated class and implicitly be fed with `self` as the first argument on their execution.

As a workaround, we should make a class that wraps this function and execute it in its `__call__` method.
Related to #1286

### PR DESCRIPTION
**Pull request by delta2323**
This PR changes `_gen_case` so that `testing.parameterize` can accept functions.

This PR fixes #1337

### CODE REVIEW

**Comment by delta2323** [comment]
In the discussion of #1337, @unnonouno and I argued the readability of the failed test cases. Here I intensionally fail the test cases by simply invoking `self.fail()` in the test fixtures. The parameters are same as the ones of `TestParameterize`.

To me, it is a bit difficult to interpret the error message when the parameter is a lambda expression.

```
======================================================================
FAIL: test_fail (test_parameterized.TestParameterize_param_0)  parameter: {'callable': <function f at 0x10be77b90>}
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/oonokenta/dev/chainer/tests/chainer_tests/testing_tests/test_parameterized.py", line 63, in test_fail
    self.fail()
AssertionError: None

======================================================================
FAIL: test_fail (test_parameterized.TestParameterize_param_1)  parameter: {'callable': <function <lambda> at 0x10be852a8>}
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/oonokenta/dev/chainer/tests/chainer_tests/testing_tests/test_parameterized.py", line 63, in test_fail
    self.fail()
AssertionError: None

======================================================================
FAIL: test_fail (test_parameterized.TestParameterize_param_2)  parameter: {'callable': <test_parameterized.C object at 0x10be65f90>}
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/oonokenta/dev/chainer/tests/chainer_tests/testing_tests/test_parameterized.py", line 63, in test_fail
    self.fail()
AssertionError: None

======================================================================
FAIL: test_fail (test_parameterized.TestParameterize_param_3)  parameter: {'callable': <bound method C.method of <test_parameterized.C object at 0x10be78690>>}
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/oonokenta/dev/chainer/tests/chainer_tests/testing_tests/test_parameterized.py", line 63, in test_fail
    self.fail()
AssertionError: None
```

**Comment by unnonouno** [comment]
Humm. How dow you think this problem?

File: tests/chainer_tests/testing_tests/test_parameterized.py [hunk 1]
```diff
@@ -34,4 +34,29 @@ def test_product_dict(self):
...
+@testing.parameterize(
+    {'callable': f},
+    {'callable': lambda x: x},
+    {'callable': C()},
+    {'callable': C().method}
+)
+class TestParameterize(unittest.TestCase):
+
+    def test_callable(self):
+        self.callable(1)
```
Comment (unnonouno) [inline tests/chainer_tests/testing_tests/test_parameterized.py]: Check if the return value is expected.

Comment (delta2323) [inline tests/chainer_tests/testing_tests/test_parameterized.py]: I fixed it. Please take a look.

**Comment by unnonouno** [comment]
The problem is not resolved, but this feature looks useful. I'll merge it.

**Pull request merged**